### PR TITLE
Add simple typing to untyped ML implementation

### PR DESCRIPTION
**Pull request by rgm**
Changes from TAPL Ch 8 -> Ch 10.

### CODE REVIEW

File: syntax.mli [hunk 1]
```diff
@@ -4,13 +4,21 @@ open Support.Pervasive
 open Support.Error
 
 (* Data type definitions *)
+type ty =
+    TyArr of ty * ty
```
Comment (rgm) [inline syntax.mli:8]: Is this supposed to be short for "arrow"?

File: core.ml [hunk 2]
```diff
@@ -6,13 +6,15 @@ open Support.Pervasive
 (* ------------------------   EVALUATION  ------------------------ *)
 
 let rec isval ctx t = match t with
-    TmAbs(_,_,_) -> true
+    TmTrue(_)  -> true
+  | TmFalse(_) -> true
```
Comment (robrix) [inline core.ml:10]: Added true and false terms

File: core.ml [hunk 3]
```diff
@@ -6,13 +6,15 @@ open Support.Pervasive
...
+  | TmAbs(_,_,_,_) -> true
```
Comment (robrix) [inline core.ml:11]: Fourth parameter for abstraction terms

File: syntax.mli (hunk 1 above)
Comment (robrix) [inline syntax.mli:8]: Yes.

File: core.ml [hunk 4]
```diff
@@ -6,13 +6,15 @@ open Support.Pervasive
...
   | _ -> false
 
 exception NoRuleApplies
 
 let rec eval1 ctx t = match t with
-    TmApp(fi,TmAbs(_,x,t12),v2) when isval ctx v2 ->
+    TmApp(fi,TmAbs(_,x,tyT11,t12),v2) when isval ctx v2 ->
```
Comment (robrix) [inline core.ml:17]: Binding the type (tho not using it). (Could use `_` to make room for it but not bind it.)

File: syntax.ml [hunk 5]
```diff
@@ -79,9 +90,8 @@ let termShift d t = termShiftAbove d 0 t
 
 let termSubst j s t =
   tmmap
-    (fun fi c x n -> if x=j+c then termShift c s else TmVar(fi,x,n))
-    0
-    t
+    (fun fi j x n -> if x=j then termShift j s else TmVar(fi,x,n))
+    j t
```
Comment (robrix) [inline syntax.ml:94]: Okay folks whatever you say.

**Pull request closed without merging**